modules/arcore/arcore_interface.cpp

```diff
@@ -0,0 +1,649 @@
+		return 0;
+	}
+}
+
+void ARCoreInterface::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("_resume"), &ARCoreInterface::_resume);
+	ClassDB::bind_method(D_METHOD("_pause"), &ARCoreInterface::_pause);
+}
+
+bool ARCoreInterface::is_initialized() const {
+	// if we're in the process of initialising we treat this as initialised...
+	return (init_status != NOT_INITIALISED) && (init_status != INITIALISE_FAILED);
+}
+
+void ARCoreInterface::_resume() {
+	if (ar_session == NULL) {
+
+		OS_Android *os_android = (OS_Android *)OS::get_singleton();
+
+		print_line("Godot ARCore: Getting environment");
+
+		// get some android things
+		JNIEnv *env = ThreadAndroid::get_env();
+
+		godot_java = os_android->get_godot_java();
+		jobject context = godot_java->get_application_context();
+		if (context == NULL) {
+			print_line("Godot ARCore: Couldn't get context");
+			init_status = INITIALISE_FAILED; // don't try again.
+			return;
+		}
+
+		print_line("Godot ARCore: Create ArSession");
+
+		// === ATTENTION!  ATTENTION!  ATTENTION! ===
+		// This method can and will fail in user-facing situations.  Your
+		// application must handle these cases at least somewhat gracefully.  See
+		// HelloAR Java sample code for reasonable behavior.
+		if (ArSession_create(env, context, &ar_session) != AR_SUCCESS || ar_session == NULL) {
+			print_line("Godot ARCore: ARCore couldn't be created.");
+			init_status = INITIALISE_FAILED; // don't try again.
+			return;
+		}
+
+		print_line("Godot ARCore: Create ArFrame.");
+
+		ArFrame_create(ar_session, &ar_frame);
+		if (ar_frame == NULL) {
+			print_line("Godot ARCore: Frame couldn't be created.");
+
+			ArSession_destroy(ar_session);
+			ar_session = NULL;
+
+			init_status = INITIALISE_FAILED; // don't try again.
+			return;
+		}
+
+		// Get our size, make sure we have these in portrait
+		Size2 size = OS::get_singleton()->get_window_size();
+		if (size.x > size.y) {
+			width = size.y;
+			height = size.x;
+		} else {
+			width = size.x;
+			height = size.y;
+		}
+
+		// Trigger display rotation
+		display_rotation = -1;
+
+		print_line("Godot ARCore: Resumed.");
+		init_status = INITIALISED;
+	}
+
+	if (init_status == INITIALISED && ar_session != NULL) {
+		ArStatus status = ArSession_resume(ar_session);
+		if (status != AR_SUCCESS) {
+			// uh, what do we do in this situation?
+		}
+	}
+}
+
+void ARCoreInterface::_pause() {
+	if (ar_session != NULL) {
+		ArSession_pause(ar_session);
+	}
+}
+
+void ARCoreInterface::notification(int p_what) {
+	// Needs testing, this should now be called
+
+	ARVRServer *arvr_server = ARVRServer::get_singleton();
+	ERR_FAIL_NULL(arvr_server);
+
+	switch (p_what) {
+		case MainLoop::NOTIFICATION_WM_FOCUS_IN: {
+			if (is_initialized()) {
+				_resume();
+
+				if (init_status == INITIALISE_FAILED) {
+					arvr_server->clear_primary_interface_if(this);
+				}
+			}
+		}; break;
+		case MainLoop::NOTIFICATION_WM_FOCUS_OUT:
+			if (is_initialized()) {
+				_pause();
+			}
+			break;
```

Comment:
New notifications were introduced that are better to detect app resumed and paused state:

- `NOTIFICATION_APP_RESUMED`: Fires when the Android activity is resumed.
- `NOTIFICATION_APP_PAUSED`: Fires when the Android activity is paused.